src/confluent_kafka/src/Producer.c

```diff
@@ -244,19 +338,30 @@ static PyObject *Producer_produce (Handle *self, PyObject *args,
 			       "partition",
 			       "callback",
 			       "on_delivery", /* Alias */
+			       "partitioner_cb",
                    "timestamp",
                    "headers",
 			       NULL };
-
 	if (!PyArg_ParseTupleAndKeywords(args, kwargs,
-					 "s|z#z#iOOLO"
+					 "s|z#z#iOOOLO"
                                          , kws,
 					 &topic, &value, &value_len,
 					 &key, &key_len, &partition,
-					 &dr_cb, &dr_cb2,
+					 &dr_cb, &dr_cb2, &partitioner_cb,
```

Comment:
The implementation of this will be somewhat similar to `stats_cb` or `logger`, not `dr_cb`.

One difference compared to any of the other callbacks is the configuration code may be more complex because we may want to be able to configure both a default custom partitioner and / or custom partitioners per topic (the .NET client allows for this). Most importantly, that allows use of the default in-built partitioner with custom partitioners on a per topic basis. To simplify things though, i think just focus on setting a global custom partitioner first up.

relevant .NET code for reference: https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/src/Confluent.Kafka/Producer.cs#L683